src/Features/CSharp/Portable/Completion/CompletionProviders/AwaitCompletionProvider.cs

```diff
@@ -0,0 +1,119 @@
+    [Shared]
+    internal sealed class AwaitCompletionProvider : LSPCompletionProvider
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public AwaitCompletionProvider()
+        {
+        }
+
+        public override ImmutableHashSet<char> TriggerCharacters => CompletionUtilities.CommonTriggerCharactersWithArgumentList;
+
+        public override async Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            var document = context.Document;
+            var position = context.Position;
+            var cancellationToken = context.CancellationToken;
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var workspace = document.Project.Solution.Workspace;
+            var syntaxContext = CSharpSyntaxContext.CreateContext(workspace, semanticModel, position, cancellationToken);
+
+            if (!syntaxContext.IsAwaitKeywordContext(position))
+            {
+                return;
+            }
+
+            var method = syntaxContext.TargetToken.GetAncestor(node => node.IsAsyncSupportingFunctionSyntax());
+            if (method is not null && !method.GetModifiers().Any(SyntaxKind.AsyncKeyword))
+            {
+                var asyncService = document.GetRequiredLanguageService<IMakeMethodAsynchronousService>();
+                var symbol = semanticModel.GetDeclaredSymbol(method, cancellationToken) as IMethodSymbol;
+                symbol ??= semanticModel.GetSymbolInfo(method, cancellationToken).Symbol as IMethodSymbol;
+                if (symbol is not null && asyncService.IsAsyncReturnType(symbol.ReturnType, new KnownTaskTypes(semanticModel.Compilation)))
+                {
+                    context.AddItem(GetCompletionItem(shouldMakeContainerAsync: true));
+                    return;
+                }
+            }
+
+            context.AddItem(GetCompletionItem(shouldMakeContainerAsync: false));
+
+            static CompletionItem GetCompletionItem(bool shouldMakeContainerAsync)
+                => CommonCompletionItem.Create(
+                       displayText: SyntaxFacts.GetText(SyntaxKind.AwaitKeyword),
+                       displayTextSuffix: "",
+                       rules: CompletionItemRules.Default,
+                       Glyph.Keyword,
+                       description: RecommendedKeyword.CreateDisplayParts(SyntaxFacts.GetText(SyntaxKind.AwaitKeyword), string.Empty),
+                       inlineDescription: shouldMakeContainerAsync ? CSharpFeaturesResources.Make_container_async : null,
+                       isComplexTextEdit: shouldMakeContainerAsync);
+        }
+
+        internal override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem completionItem, TextSpan completionListSpan, char? commitKey, bool disallowAddingImports, CancellationToken cancellationToken)
```

Comment:
One thing I would ask is that we rationalize the public/internal GetChangeAsync here so we don't have to expose stuff through ExternalAccess.

Reply:
@333fred Thanks for noting. I actually forgot that I have this in my PR when I was looking at your PR :smile:

I was using the `completionListSpan` argument that doesn't exist for the public overload. I'll try to use `completionItem.Span` instead and see if things still work.

Reply:
It worked! Switched to the public overload. Thanks @333fred.

Reply:
@CyrusNajmabadi was going to noodle on the general structure of hierarchy as well, so it was more a comment to make sure we include this in whatever decision we end up making for the whole thing.